### PR TITLE
WIP: First draft of DeFleCT module

### PR DESCRIPTION
**Pull request by olafhauk**
If you want to feel productive - I'm sure there's still lots to do here. I would be very happy about any feedback.

### CODE REVIEW

**Comment by wmvanvliet** [comment]
Interesting. I've just read through your DeFleCT paper, which places two of MNE's source estimation algorithms (MNE, LCMV beamformers) in the same framework, allowing us to mix and match the properties of both. I particularly like the ability to explicitly 'shut up' certain regions which you know are not of interest to you but do leak into the regions you are interested in. Eye movements came to mind... (estimating the topographies of those is not that difficult).

**Some stuff to do:**

The example in its current form would fit better as a tutorial in the new-style manual we are developing. The examples should mainly demonstrate how the API words and preferably make the user go "wow, that's so easy!". Why not show two estimators in the example and skip the connectivity analysis?

The API needs a close look. Private functions should start with an underscore. Docstrings should be in the proper format.

File: mne/minimum_norm/DeFleCT.py [hunk 1]
```diff
@@ -0,0 +1,463 @@
...
+import scipy
+from mne.utils import logger, verbose
+from mne import pick_types, pick_channels_forward, pick_channels_cov, SourceEstimate
+from mne.cov import prepare_noise_cov 
+from mne.cov import regularize as cov_regularize
+from mne.proj import make_eeg_average_ref_proj 
+from mne.minimum_norm.inverse import _prepare_forward
+
+
+def DeFleCT_matrix(F, P, i, t, lambda2_S=1/9.):
```
Comment (wmvanvliet) [inline mne/minimum_norm/DeFleCT.py]: private functions should start with underscore. Maybe also just make it lower case: `_deflect_matrix`

File: mne/minimum_norm/DeFleCT.py [hunk 2]
```diff
@@ -0,0 +1,463 @@
...
+    """ Computes the linear DeFleCT estimator given its component matrices
+
+    Parameters:
+    -----------
+    F: leadfield/forward solution matrix (n_chan x n_vert)
```
Comment (wmvanvliet) [inline mne/minimum_norm/DeFleCT.py]: Parameter documentation follows a strict syntax:

    F : 2D array (n_chan x n_vert)
        Leadfield / forward solution matrix

File: mne/minimum_norm/DeFleCT.py [hunk 3]
```diff
@@ -0,0 +1,463 @@
...
+
+    # w = (tF^t + (i - tF^t S-1 P) (P^t S-1 P)-1 P^t) S-1  
+    # S = F F^t + lambda2 C
+
+    # for whitened data, covariance matrix is identity
+
+    [n_chan, n_comp] = P.shape
+    [n_chan2, n_vert] = F.shape
+    if n_chan != n_chan2:
+        raise ValueError("\nNumber of channels doesn't match between P and F: \
```
Comment (wmvanvliet) [inline mne/minimum_norm/DeFleCT.py]: why the initial `\n` in the error message string?

File: mne/minimum_norm/DeFleCT.py [hunk 4]
```diff
@@ -0,0 +1,463 @@
...
+                            %d vs %d" % (n_chan, n_chan2))
```
Comment (wmvanvliet) [inline mne/minimum_norm/DeFleCT.py]: MNE's coding style avoids the use of `\` to break lines. Use string comprehension instead:

    raise ValueError("Number of channels doesn't match between P and F: "
                     "%d vs %d" % (n_chan, n_chan2)))

File: mne/minimum_norm/DeFleCT.py [hunk 5]
```diff
@@ -0,0 +1,463 @@
...
+                                           this_label_lfd_summary, 1)
+
+    return label_lfd_summary
+
+# Done
+
+
+def DeFleCT_make_estimator(forward, noise_cov, labels, lambda2_cov=3/10.,
+                           lambda2_S=1/9., pick_meg=True, pick_eeg=False,
+                           mode='svd', n_svd_comp=1, verbose=None):
```
Comment (wmvanvliet) [inline mne/minimum_norm/DeFleCT.py]: Right now, calling the API for computing inverse operators looks like this:

    mne.minimum_norm.make_inverse_operator   # MNE
    mne.minimum_norm.DeFleCT_make_estimator  # DeFleCT

maybe we should change things around to this:

    mne.minimum_norm.make_inverse_operator   # MNE
    mne.deflect.make_inverse_operator  # DeFleCT

**Comment by wmvanvliet** [comment]
can't wait to play with this. Thanks for opening this PR Olaf!

**Comment by olafhauk** [comment]
Thanks for your suggestions - I've just been away for a while, but will take of this in the next couple of weeks.

File: mne/minimum_norm/DeFleCT.py (hunk 1 above)
Comment (olafhauk) [inline mne/minimum_norm/DeFleCT.py]: Done (only for private functions). Some MNE functions do capitalize (e.g. SourceEstimate).

File: mne/minimum_norm/DeFleCT.py (hunk 2 above)
Comment (olafhauk) [inline mne/minimum_norm/DeFleCT.py]: Done to the best of my capabilities.

File: mne/minimum_norm/DeFleCT.py (hunk 3 above)
Comment (olafhauk) [inline mne/minimum_norm/DeFleCT.py]: Removed. Although I find screen output easier to follow if there are more empty lines.

File: mne/minimum_norm/DeFleCT.py (hunk 4 above)
Comment (olafhauk) [inline mne/minimum_norm/DeFleCT.py]: Done.

File: mne/minimum_norm/DeFleCT.py (hunk 5 above)
Comment (olafhauk) [inline mne/minimum_norm/DeFleCT.py]: I don't have a strong opinion on this - but I think you are right, DeFleCT isn't a minimum norm method. But that's also where my "psf_ctf" functions are. I have not changed this yet.

File: mne/minimum_norm/DeFleCT.py (hunk 1 above)
Comment (agramfort) [inline mne/minimum_norm/DeFleCT.py]: SourceEstimate is not a function but a class. That's expected to capitalize a class but not for
functions.

**Comment by agramfort** [comment]
@olafhauk I think you did not push your recent commits.

**Comment by olafhauk** [comment]
I've got problems pushing my latest commits. I get the message:

error: The requested URL returned error: 403 Forbidden while accessing https://github.com/mne-tools/mne-python.git/info/refs?service=git-receive-pack
fatal: HTTP request failed

I've tried https and ssh, but neither works. I can clone mne-python, i.e. it's not a general connection problem. I've spent 30 minutes with our IT boss looking at this. 

Any ideas what I could try next? It worked last time.

**Comment by olafhauk** [comment]
Ok - I think it worked now...

File: mne/minimum_norm/DeFleCT.py [hunk 6]
```diff
@@ -0,0 +1,482 @@
+# Authors: Olaf Hauk <olaf.hauk@mrc-cbu.cam.ac.uk>
```
Comment (agramfort) [inline mne/minimum_norm/DeFleCT.py]: rename the file in small letters please.

File: examples/plot_deflect_spatial_filter.py [hunk 7]
```diff
@@ -0,0 +1,236 @@
...
+
+The constraints that can be implemented are:
+- Suppress cross-talk within a small set of labels
+- Minimize cross-talk from other locations anywhere in the brain
+- Minimize the effect of noise (represented by noise covariance matrix).
+
+Different sensor types are combined by whitening with noise covariance matrix.
+First label is the target. Cross-talk with other labels should be zero. 
+Cross-talk with other vertices will be minimized.
+For reference, see http://www.ncbi.nlm.nih.gov/pubmed/23616402
```
Comment (agramfort) [inline examples/plot_deflect_spatial_filter.py]: you're example is too long. The more concise the easier it is to reuse.

I would just stick to the MEG data.

**Comment by agramfort** [comment]
you'll need to add tests and investigate why travis is not happy.

thanks @olafhauk

**Pull request closed without merging**